### PR TITLE
Symmetric quantized convolution kernel ARM64

### PR DESCRIPTION
**Pull request by chenfucn**
**Description**: Describe your changes.
Adding a symmetric quantized convolution kernel for ARM64

Pre-dot CPU (Pixel 3a)
| Model                 | baseline    | this change |
| ------------------ | ----------- |  ----------- | 
| Cartoongan        | 1543.21    |  1530.02 |
| Efficient Net       | 103.886    |  100.332 |
| Mobilenet Edge TPU | 33.3259 | 33.0789 |
| Mobilenet v1  | 19.177 | 19.995 |

Post-dot CPU (Pixel 5a)
| Model                 | baseline    | this change |
| ------------------ | ----------- |  ----------- | 
| Cartoongan        | 734.349    |  708.554 |
| Efficient Net       | 57.5911    |  56.9976 |
| Mobilenet Edge TPU | 11.3029 | 10.7636 |
| Mobilenet v1  | 8.24782 | 8.05804 |

Note:
Indirect conv performs worse for shallow convs (input channels are small).  This is much more so for low end pre-dot CPUs, where only 128 or deeper conv is faster with indirect conv. With DOT-CPUs, 32 deep conv is already faster

### CODE REVIEW

File: onnxruntime/core/mlas/lib/convsym.cpp [hunk 1]
```diff
@@ -232,6 +233,12 @@ MlasConvSymPackWSize(
 
     } else {
 
+#ifdef MLAS_TARGET_ARM64
+        if (InputChannels < 32) {
+            return 0;
+        }
+#endif
+
         size_t OutputChannelPackCount = ConvSymDispatch->FilterOutputChannelPackCount;
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/convsym.cpp:242]: another case is we should not use this kernel on device with dot product

File: onnxruntime/test/providers/cpu/nn/qlinearconv_op_test.cc [hunk 2]
```diff
@@ -581,6 +581,36 @@ TEST(QLinearConvTest, Conv1D_U8S8) {
   test.Run();
 }
 
+TEST(QLinearConvTest, Conv2D_U8S8aaa) {
```
Comment (yufenglee) [inline onnxruntime/test/providers/cpu/nn/qlinearconv_op_test.cc:584]: nit: a better name

File: onnxruntime/core/mlas/lib/mlasi.h [hunk 3]
```diff
@@ -103,7 +103,41 @@ Module Name:
...
+   private:
+    MLASCPUIDInfo();
+
+    bool has_arm_neon_dot_{false};
+
+    static MLASCPUIDInfo instance_;
+};
+using MLAS_CPUIDINFO = MLASCPUIDInfo;
+
+#endif // MLAS_TARGET_ARM64
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/mlasi.h:138]: why do you put it under MLAS_NO_ONNXRUNTIME_THREADPOOL check?

File: onnxruntime/core/mlas/lib/mlasi.h [hunk 4]
```diff
@@ -103,7 +103,41 @@ Module Name:
...
+
+    // ARM
+    bool HasArmNeonDot() const { return has_arm_neon_dot_; }
+
+   private:
+    MLASCPUIDInfo();
+
+    bool has_arm_neon_dot_{false};
+
+    static MLASCPUIDInfo instance_;
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/mlasi.h:134]: >static MLASCPUIDInfo instance_; [](http://example.com/codeflow?start=4&length=31)

a dup of 'static MLASCPUIDInfo cpuid_info;' in factory fuction?

File: onnxruntime/core/mlas/lib/mlasi.h (hunk 3 above)
Comment (yufenglee) [reply onnxruntime/core/mlas/lib/mlasi.h:138]: It is better to give it a more general name for MLAS_NO_ONNXRUNTIME_THREADPOOL.

File: onnxruntime/test/providers/cpu/nn/qlinearconv_op_test.cc [hunk 5]
```diff
@@ -595,16 +595,6 @@ TEST(QLinearConvTest, Conv2D_U8S8_Sym_M64_C64) {
   test.Run();
 }
 
-TEST(QLinearConvTest, Conv2D_U8S8_Sym_M16_C4) {
```
Comment (yufenglee) [inline onnxruntime/test/providers/cpu/nn/qlinearconv_op_test.cc:598]: keep this unit test?

File: onnxruntime/core/mlas/lib/aarch64/ConvSymU8KernelNeon.S [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+        b.hi    .LConvSym.KernelSizeLoop
+        b       .LConvSym.Requantize
+
+.LConvSym.PartialStore:
+        tbz     x6,2,7f
+        st1     {v0.s}[2],[x5],4
+        str     s0,[x2],4
+        EXT     v0.16b,v0.16b,v0.16b,4
+
+7:
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/aarch64/ConvSymU8KernelNeon.S:443]: nit: 1f or .LConvSym.PartialStore2, in which 2 means storing 2 items.

File: onnxruntime/test/providers/cpu/nn/qlinearconv_op_test.cc (hunk 5 above)
Comment (chenfucn) [reply onnxruntime/test/providers/cpu/nn/qlinearconv_op_test.cc:598]: same with Conv2D_U8S8_Sym_M16_C4_Bias

File: onnxruntime/core/mlas/lib/aarch64/ConvSymU8KernelDot.S [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+#include "AssembleDotProduct.h"
+
+        .equ    .LMLAS_CONV_SYM_FLAG_INPUT_DIRECT,      1
+        .equ    .LMLAS_CONV_SYM_FLAG_PER_CHANNEL_SCALE, 2
+
+//
+// Stack frame layout for the symmetric convolution kernel.
+// d8-d15, x19-x30 need to be preserved if used
+//
+        .equ    .LConvSymFrame_SavedRegisters,       (4 * 8)
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/aarch64/ConvSymU8KernelDot.S:28]: >4 [](http://example.com/codeflow?start=54&length=1)

nit: 3*8 is enough

**Review by yufenglee — DISMISSED** [review]
:shipit:

File: onnxruntime/core/mlas/lib/arm64/ConvSymU8KernelDot.asm [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+        sdot    v24.4s,v6.16b,v8.4b[1]
+        sdot    v25.4s,v6.16b,v9.4b[1]
+        sdot    v26.4s,v6.16b,v10.4b[1]
+        sdot    v27.4s,v6.16b,v11.4b[1]
+        sdot    v28.4s,v7.16b,v8.4b[1]
+        sdot    v29.4s,v7.16b,v9.4b[1]
+        sdot    v30.4s,v7.16b,v10.4b[1]
+        sdot    v31.4s,v7.16b,v11.4b[1]
+
+        TST     x7,15
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/arm64/ConvSymU8KernelDot.asm:388]: >TST [](http://example.com/codeflow?start=8&length=3)

Please keep it as lower case and for the instructions below.

**Review by yufenglee — APPROVED** [review]
:shipit:

**Pull request merged**